Nightwatch/Service.fs

```diff
@@ -0,0 +1,118 @@
+open Nightwatch.ResourceConfiguration
+open Nightwatch.Resources
+
+type ProgramInfo =
+    { version : Version }
+
+let private logFullVersion (logger : ILogger) programInfo =
+    logger.Information("Nightwatch v. {0}", programInfo.version)
+    logger.Information("Config file format v. {0}", ResourceConfiguration.configFormatVersion)
+
+let private resourceFactories = [| Http.factory Http.system; Shell.factory Process.system |]
+
+let private configureResourceRegistry (logger : ILogger) =
+    let names = resourceFactories |> Seq.map (fun f -> f.resourceType)
+    logger.Information("Available resources: {0}", String.Join(", ", names))
+    Registry.create resourceFactories
+
+let private readResources fs config factories : Async<Result<Resource [], InvalidConfiguration []>> =
+    let splitResults seq =
+        let chooseOk = function
+        | Ok x -> Some x
+        | Error _ -> None
+
+        let chooseError = function
+        | Ok _ -> None
+        | Error x -> Some x
+
+        let isOk = chooseOk >> Option.isSome
+
+        let (results, errors) =
+            seq
+            |> Seq.toArray
+            |> Array.partition isOk
+        Array.choose chooseOk results, Array.choose chooseError errors
+
+    async {
+        let! resources = Async.AwaitTask <| ResourceConfiguration.read factories fs config
+        let (results, errors) = splitResults resources
+        return
+            if Seq.isEmpty errors
+            then Ok results
+            else Error errors
+    }
+
+let private createScheduler resources =
+    let schedule = Scheduler.prepareSchedule resources
+    task {
+        let! scheduler = Scheduler.create()
+        do! Scheduler.configure scheduler schedule
+        return scheduler
+    }
+
+let startService logger programInfo env fs (configFilePath : Path) =
+    let errorsToString errors =
+        let printError { path = (Path path); id = id; message = message } =
+            sprintf "Path: %s\nId: %s\nMessage: %s"
+                path
+                (defaultArg id "N/A")
+                message
+
+        String.Join("\n", Seq.map printError errors)
+
+    async {
+        logFullVersion logger programInfo
+        logger.Information("Configuration file location: {0}", configFilePath)
+        let! config = ProgramConfiguration.read env fs configFilePath
+        logger.Information("Resource directory location: {0}", config.resourceDirectory)
+        let registry = configureResourceRegistry logger
+        let! resources = readResources fs config registry
+        match resources with
+        | Ok resources ->
+            logger.Information("{0} resources loaded", resources.Length)
+            let! scheduler = Async.AwaitTask <| createScheduler resources
+            do! Scheduler.start scheduler
+            return Some scheduler
+        | Error errors ->
+            logger.Error(errorsToString errors)
+            return None
+    }
+
+type HostedService(logger : ILogger,

```

Comment:
I'd prefer dumb type parametrized by function `unit -> Async<Scheduler>` instead of explicit dependencies.

Reply:
That way, I'll have to make my messy `startService` signature public, and that's opposing to what I tried to do with this PR. Servicing stuff has to be encapsulated in the module.

And, ideally, this class should be serviceable with ordinary DI structures (it doesn't work that way yet because of `Path` dependency, but I think it's good enough already).

The purpose of this PR was to cleanly (I mean cleanly on .NET Core DI terms) integrate our stuff with Generic Host API.

Reply:
I see. Thanks.